GG/src/GUI.cpp

```diff
@@ -314,38 +322,44 @@ void GUIImpl::HandleMouseButtonPress(unsigned int mouse_button, const Pt& pos, i
     m_prev_mouse_button_press_pos = pos;
 
     m_mouse_button_state[mouse_button] = true;
-    m_drag_wnds[mouse_button] = m_curr_wnd_under_cursor; // track this window as the one being dragged by this mouse button
-    if (m_curr_wnd_under_cursor) {
-        m_prev_wnd_drag_position = m_curr_wnd_under_cursor->UpperLeft();
-        m_wnd_drag_offset = pos - m_prev_wnd_drag_position;
+
+    // If not already dragging, start tracking the dragged window.
+    if (m_drag_wnds[0].expired() && m_drag_wnds[1].expired() && m_drag_wnds[2].expired()) {
+        m_drag_wnds[mouse_button] = curr_wnd_under_cursor;
+        if (curr_wnd_under_cursor) {
+            m_prev_wnd_drag_position = curr_wnd_under_cursor->UpperLeft();
+            m_wnd_drag_offset = pos - m_prev_wnd_drag_position;
+        }
     }
 
     // if this window is not a disabled Control window, it becomes the focus window
-    Control* control = nullptr;
-    if (m_curr_wnd_under_cursor && (!(control = dynamic_cast<Control*>(m_curr_wnd_under_cursor)) || !control->Disabled()))
-        GUI::s_gui->SetFocusWnd(m_curr_wnd_under_cursor);
-
-    if (m_curr_wnd_under_cursor) {
-        m_wnd_region = m_curr_wnd_under_cursor->WindowRegion(pos); // and determine whether a resize-region of it is being dragged
+    auto control = (curr_wnd_under_cursor
+                    ? dynamic_cast<Control*>(curr_wnd_under_cursor.get())
```

Comment:
Why get a raw pointer and not `dynamic_pointer_cast` on `curr_wnd_under_cursor` (or `m_curr_wnd_under_cursor`)?